### PR TITLE
[Doc] Add EN custmized runtime doc in dev-1.x

### PR DESCRIPTION
**Pull request by MengzhangLI**
## Motivation
Translate Chinese version customized runtime doc into English https://github.com/open-mmlab/mmsegmentation/pull/2502.

### CODE REVIEW

File: docs/en/advanced_guides/customize_runtime.md [hunk 1]
```diff
@@ -1,245 +1,168 @@
 # Customize Runtime Settings
 
-## Customize optimization settings
+## Customize hooks
 
-### Customize optimizer supported by Pytorch
+### Step 1: Create a new hook
```
Comment (MeowZheng) [inline docs/en/advanced_guides/customize_runtime.md:5]:
```suggestion
### Step 1: Implement a new hook
```

File: docs/en/advanced_guides/customize_runtime.md [hunk 2]
```diff
@@ -1,245 +1,168 @@
...
 
-We already support to use all the optimizers implemented by PyTorch, and the only modification is to change the `optimizer` field of config files.
-For example, if you want to use `ADAM` (note that the performance could drop a lot), the modification could be as the following.
+MMEngine has implemented commonly used [hooks](https://github.com/open-mmlab/mmengine/blob/main/docs/en/tutorials/hook.md) for training and test,
+When users have requirements for customization, they can follow examples below.
+For example, if they want to modify the value of hyperparameter `model.hyper_paramete`, aiming at making it change with training iteration number:
```
Comment (MeowZheng) [inline docs/en/advanced_guides/customize_runtime.md:9]:
```suggestion
For example, if some hyper-parameter of the model needs to be changed when model training, we can implement a new hook for it:
```

File: docs/en/advanced_guides/customize_runtime.md [hunk 3]
```diff
@@ -1,245 +1,168 @@
...
+        model.hyper_parameter = self.a * cur_iter + self.b
 ```
 
-#### 2. Add the optimizer to registry
-
-To find the above module defined above, this module should be imported into the main namespace at first. There are two options to achieve it.
+### Step 2: Import a new hook
 
-- Modify `mmseg/core/optimizer/__init__.py` to import it.
+The module which is defined above needs to be imported into main namespace first to ensure being detected by program.
```
Comment (MeowZheng) [inline docs/en/advanced_guides/customize_runtime.md:43]:
```suggestion
The module which is defined above needs to be imported into main namespace first to ensure being registered.
```

File: docs/en/advanced_guides/customize_runtime.md [hunk 4]
```diff
@@ -1,245 +1,168 @@
...
+Assume NewHook is in `mmseg/engine/hooks/new_hook.py`, there are two ways to implement it:
```
Comment (MeowZheng) [inline docs/en/advanced_guides/customize_runtime.md:44]:
```suggestion
We assume `NewHook` is implemented in `mmseg/engine/hooks/new_hook.py`, there are two ways to import it:
```

File: docs/en/advanced_guides/customize_runtime.md [hunk 5]
```diff
@@ -1,245 +1,168 @@
...
-
-Actually users can use a totally different file directory structure using this importing method, as long as the module root can be located in `PYTHONPATH`.
-
-#### 3. Specify the optimizer in the config file
+### Step 3: Modify config file
 
-Then you can use `MyOptimizer` in `optimizer` field of config files.
-In the configs, the optimizers are defined by the field `optimizer` like the following:
+Users can set and use customized hooks in training and test followed methods below.
+The priorities of different hooks which are located in the same place can be referred [here](https://github.com/open-mmlab/mmengine/blob/main/docs/en/tutorials/hook.md#built-in-hooks),
```
Comment (MeowZheng) [inline docs/en/advanced_guides/customize_runtime.md:64]:
```suggestion
The execution priority of hooks at the same place of `Runner` can be referred [here](https://github.com/open-mmlab/mmengine/blob/main/docs/en/tutorials/hook.md#built-in-hooks),
```

File: docs/en/advanced_guides/customize_runtime.md [hunk 6]
```diff
@@ -1,245 +1,168 @@
...
 
-To use your own optimizer, the field can be changed to
-
-```python
-optimizer = dict(type='MyOptimizer', a=a_value, b=b_value, c=c_value)
-```
+## Customize optimizer
 
-### Customize optimizer constructor
+### Step 1: Create a new optimizer
```
Comment (MeowZheng) [inline docs/en/advanced_guides/customize_runtime.md:75]:
```suggestion
### Step 1: Implement a new optimizer
```

File: docs/en/advanced_guides/customize_runtime.md [hunk 7]
```diff
@@ -1,245 +1,168 @@
...
 
-Some models may have some parameter-specific settings for optimization, e.g. weight decay for BatchNorm layers.
-The users can do those fine-grained parameter tuning through customizing optimizer constructor.
+If users want to add a new optimizer `MyOptimizer` which has parameters `a`, `b` and `c`. We recommend implementing it in `mmseg/engine/optimizers/my_optimizer.py`:
```
Comment (MeowZheng) [inline docs/en/advanced_guides/customize_runtime.md:77]:
```suggestion
We recommend the customized optimizer implemented in `mmseg/engine/optimizers/my_optimizer.py`. Here is an example of a new optimizer `MyOptimizer` which has parameters `a`, `b` and `c`:
```

File: docs/en/advanced_guides/customize_runtime.md [hunk 8]
```diff
@@ -1,245 +1,168 @@
...
-      warmup='linear',
-      warmup_iters=1000,
-      warmup_ratio=1.0 / 10,
-      min_lr_ratio=1e-5)
-  ```
+### Step 2: Import a new optimizer
 
-## Customize workflow
+The module which is defined above needs to be imported into main namespace first to ensure being detected by program.
+Assume `MyOptimizer` is in `mmseg/engine/optimizers/my_optimizer.py`, there are two ways to implement it:
```
Comment (MeowZheng) [inline docs/en/advanced_guides/customize_runtime.md:93]: same  problem as above

File: docs/en/advanced_guides/customize_runtime.md [hunk 9]
```diff
@@ -1,245 +1,168 @@
...
+                     optimizer=dict(type='MyOptimizer',
+                                    a=a_value, b=b_value, c=c_value),
+                     clip_grad=None)
+```
 
-## Customize hooks
+## Customize optimizer constructor
 
-### Use hooks implemented in MMCV
+### Step 1: Create a new optimizer constructor
```
Comment (MeowZheng) [inline docs/en/advanced_guides/customize_runtime.md:121]:
```suggestion
### Step 1: Implement a new optimizer constructor
```

File: docs/en/advanced_guides/customize_runtime.md [hunk 10]
```diff
@@ -1,245 +1,168 @@
...
 
-If the hook is already implemented in MMCV, you can directly modify the config to use the hook as below
+Constructor can be used to create optimizer, optimizer wrapper and customized hyperparameters of different layers in model network. The optimizer of some models would adjust with specified parameters such as weight decay of BatchNorm layer.
+Users can set different optimization policy on different parameters of model by customized optimizer constructor.
```
Comment (MeowZheng) [inline docs/en/advanced_guides/customize_runtime.md:124]:
```suggestion
Optimizer constructor is used to create optimizer and optimizer wrapper for model training, which has powerful functions like specifying learning rate and weight decay for different model layers. Here is an example for a customized optimizer constructor.
```

File: docs/en/advanced_guides/customize_runtime.md [hunk 11]
```diff
@@ -1,245 +1,168 @@
...
-- optimizer_config
-- momentum_config
+        return my_optimizer
+```
 
-In those hooks, only the logger hook has the `VERY_LOW` priority, others' priority are `NORMAL`.
-The above-mentioned tutorials already covers how to modify `optimizer_config`, `momentum_config`, and `lr_config`.
-Here we reveals how what we can do with `log_config`, `checkpoint_config`, and `evaluation`.
+Default optimizer constructor is implemented [here](https://github.com/open-mmlab/mmengine/blob/main/mmengine/optim/optimizer/default_constructor.py#L19).
+It can also be used as template of new optimizer constructor.
```
Comment (MeowZheng) [inline docs/en/advanced_guides/customize_runtime.md:140]:
```suggestion
It can also be used as base class of new optimizer constructor.
```

File: docs/en/advanced_guides/customize_runtime.md [hunk 12]
```diff
@@ -1,245 +1,168 @@
...
 
-#### Checkpoint config
+### Step 2: Import a new optimizer constructor
 
-The MMCV runner will use `checkpoint_config` to initialize [`CheckpointHook`](https://github.com/open-mmlab/mmcv/blob/9ecd6b0d5ff9d2172c49a182eaa669e9f27bb8e7/mmcv/runner/hooks/checkpoint.py#L9).
+The module which is defined above needs to be imported into main namespace first to ensure being detected by program.
+Assume `MyOptimizerConstructor` is in `mmseg/engine/optimizers/my_optimizer_constructor.py`, there are two ways to implement it:
```
Comment (MeowZheng) [inline docs/en/advanced_guides/customize_runtime.md:145]: same problem as above

File: docs/en/advanced_guides/customize_runtime.md [hunk 13]
```diff
@@ -1,245 +1,168 @@
...
+    def __init__(self, a: int, b: int) -> None:
+        self.a = a
+        self.b = b
 
+    def before_train_iter(self,
+                          runner,
+                          batch_idx: int,
+                          data_batch: Optional[Sequence[dict]] = None) -> None:
+        cur_iter = runner.iter
+        # 当模型被包在 wrapper 里时获取这个模型
```
Comment (MeowZheng) [inline docs/en/advanced_guides/customize_runtime.md:35]: translate in English

**Review by MeowZheng — APPROVED** [review]
